feat: v1

tranhl opened the pull request:
## Stack

<!-- branch-stack -->

- `main`
  - \#1 :point\_left:

kevgo commented on README.md:
```diff
@@ -1,2 +1,156 @@
+
+# Git Town Action V1
+
+This action synchronizes visualizes your stacked changes when proposing pull requests on GitHub. This allows you to easily see all related PRs for a given pull request, where
```
```suggestion
This action visualizes your stacked changes when proposing pull requests on GitHub. This allows you to easily see all related PRs for a given pull request, where
```

kevgo commented on DEVELOPMENT.md:
```diff
@@ -0,0 +1,9 @@
+# Developing the Git Town action source code
+
+This page helps you get started hacking on the Git Town codebase.
```
```suggestion
This page helps you get started hacking on the Git Town GitHub Action codebase.
```

kevgo commented on README.md:
```diff
@@ -1,2 +1,156 @@
+    # The regex used by the action to find and replace the visualization
+    # within a PR's body.
+    # 
+    # Default: '[ \t]*?## Stack[\s\S]*?(?=##|$)'
```
Make the number of `#` flexible here?

tranhl replied:
Ah yes that didn't even cross my mind!

I'll have to circle back to this last though, as I gave it a crack and couldn't get the lookahead for the next section (`(?=##|$)`) to be flexible as well.

kevgo replied:
I agree this can get tricky. Considering that the next section might not use the same heading level (amount of `#`) as this section, it might be enough to just match for  `^#+`, i.e. any amount of `#` at the beginning of a line.

kevgo commented on README.md:
```diff
@@ -1,2 +1,156 @@
+      {{ stack }}
+
+    # Update the regex to look for this header
+    body-regex: '[ \t]*?# Stacked Changes[\s\S]*?(?=#|$)'
```
Question around this regex: AFAIK `[\s\S]` includes both whitespace and non-whitespace characters, i.e. matches all characters. Could it be replaced with `.`?

tranhl replied:
Good question! Unfortunately not as `.` does not match newline characters, meaning the match would end after the first line.

kevgo replied:
Putting the regex in [dotall mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/dotAll) might help makes the `.` match newlines.

kevgo commented on DEVELOPMENT.md:
```diff
@@ -0,0 +1,9 @@
+
+## setup
+
+1. install Node.js v18+
```
In `package.json` it runs esbuild with `--target=node20`. Would it make sense to update the documented Node version to 20 here?

tranhl replied:
Yep definitely, good catch! I'll include it in [package.json.engines](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#engines) as well to help enforce that consistency.

kevgo commented on package.json:
```diff
@@ -0,0 +1,30 @@
+  "main": "dist/index.js",
+  "scripts": {
+    "dev": "npm run build --watch",
+    "build": "esbuild src/index.ts --outfile=dist/index.js --bundle --minify --platform=node --target=node20",
```
+1 for using `esbuild`! :rocket:

kevgo commented on src/inputs.ts:
```diff
@@ -0,0 +1,119 @@
+  getCurrentPullRequest(context: typeof github.context) {
+    try {
+      const pullRequest:
+        | {
+            number: number
+            base?: { ref?: string }
+            head?: { ref?: string }
+          }
```
Extract this type into types.ts?

tranhl replied:
The type is purely for working with `base` and `head`, which are both `any` due to `context.payload.pull_request` having the following type:

```ts
type WebhookPayload = {
  pull_request: {
    [key: string]: any;
    number: number;
    html_url?: string | undefined;
    body?: string | undefined;
  } | undefined
}
```

Since `inputs.ts` abstracts this away, I don't believe there's much value in extracting this type out, since (at least I hope) it wouldn't be used elsewhere.

kevgo replied:
Cool, thanks for the background!

kevgo reviewed (dismissed):
Thanks for this, very nicely done :pray: I just left a few suggestions and comments, no showstoppers. Feel free to adopt or disregard them and merge this yourself when you are ready so that you can make sure it works. I'm excited that this is open-source now!

tranhl commented:
@kevgo Oh hey thanks for taking the time to review it! This was meant to be a draft PR as I still need to iron out the CI/CD and automation for the action. Nevertheless thanks for the review!

kevgo reviewed (commented):
Sorry I'm so used to reviewing everything that gets submitted to Git Town that I just jumped on this PR. I'll only engage when you tag me in the future. Thanks again!

tranhl commented:
> Sorry I'm so used to reviewing everything that gets submitted to Git Town that I just jumped on this PR. I'll only engage when you tag me in the future. Thanks again!

No worries! I got some useful feedback thanks to it!

> P.S. I'll need your help enabling actions on this repository. They're currently disabled, preventing me from testing out the CI/CD setup. Thank you!
Did you happen to catch this? Still need your help on this one so that I can test the CI/CD workflows for this repository. Thanks 🙏

kevgo commented:
>> P.S. I'll need your help enabling actions on this repository. They're currently disabled, preventing me from testing out the CI/CD setup. Thank you!

> Did you happen to catch this? Still need your help on this one so that I can test the CI/CD workflows for this repository. Thanks 🙏

Sorry I had missed that part and thanks for following up on this. The org settings allow all Actions now. I think I was playing with those settings a few years ago and then forgot about them. This should work for you now. Apologies for unnecessary slowing you down and please reach out if you need anything else!

tranhl commented:
@kevgo It's been a while! Apologies for the wait, but the action is finally in a state I'm happy with! I've since added support for the `perennial-regex` config option, but more importantly, I've removed the `body-regex` input from this action.

I didn't like using regex to configure what sections of the PR body would be replaced. It was too clunky, and lead to poor UX since users would have to figure out a nasty regex to get it to work 🥲. We still needed the user to tell the action what to replace in the body though. Turns out, the offical GitHub actions had a similar problem, and they solved it by using [HTML comment tags](https://github.com/actions/checkout/blob/cd7d8d697e10461458bc61a30d094dc601a8b017/README.md?plain=1#L19) to indicate the start/end of templated sections in Markdown. I adopted that solution for this action, so now users just need to add `<!-- stack-start -->` and `<!-- stack-end -->` to their PR template, and a visualisation will be inserted in between! I think is much easier to grok, and more reliable at the end of the day as it doesn't rely on regex black magic. I'd love to get your thoughts on this!

Lastly, it would be great if you could give everything one last look over before I merge this in. Super excited to finally ship this and use it!

kevgo commented on README.md:
````diff
@@ -1,2 +1,185 @@
+
+```yaml
+- uses: git-town/action@v1
+  with:
+    template: |
+      # Stacked Changes
+      {{ stack }}
````
I have never used this action, so I might be unaware of usability issues here. You certainly have more experience with this and I trust your judgement. With this said, I wonder whether we really need two separate templating systems here:

1. The `<!-- stack-start -->` and `<!-- stack-end -->` placeholders in the pull request description.
2. The template here that gets inserted between these two tags and uses `{{ stack }}` as a placeholder.

Having two separate placeholders and placeholder formats seems like a lot for such a simple feature. Could we leave templating system 2 out? If the user wants to have a `# Stacked Changes` header above their stacked change in their PR desciption, they can add it to their PR description. Teams can use [pull-request templates](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) if they want the pull request descriptions (including the stack visualization in them) to look in a particular way.

tranhl replied:
I can actually answer this question and the previous question around using two anchor tags in one go! The reason is  that if users don't use a PR template, then the visualisation will be added to the PR without a heading! However, thinking through this now, I think this isn't too big of a deal, as the majority of people interested in this action will probably be worked in teams. The chances that teams aren't using a pull request template would be pretty low!

I'll go ahead and make a few changes here:

1. Remove the need to have two anchor tags. We'll use a single tag `<!-- branch-stack -->`, and during updates, we'll look for a list node immediately following the tag and replace everything in between with the updated visualisation.
2. Remove the template input. This will give users the freedom to structure their PR body any way they want. If the tag isn't present, then we'll simply append it to the bottom of the PR body.

kevgo replied:
I like your new approach!

kevgo reviewed (commented):
This is starting to look really cool, thanks for the great work on this! 🚀 Just a few observations and comments below.

tranhl commented on .git-branches.toml:
```diff
@@ -0,0 +1,17 @@
+main = "main"
+
+[hosting]
+platform = "github"
```
It doesn't, I've simply included it as part of being explicit over implicit! Happy to remove it if we prefer for Git Town to auto-detect for this repo.

kevgo replied:
Being explicit is a good point. I just wanted to point out the ability for auto-detect. Your call.

tranhl commented on .github/workflows/check-dist.yml:
```diff
@@ -0,0 +1,51 @@
+      - uses: actions/checkout@v3
+
+      - name: Set Node.js 20.x
+        uses: actions/setup-node@v1
```
Yep for sure, good catch! We bootstrapped the action from an action template that itself was bootstrapped from an older version of GitHub's TypeScript action template, so things were out of date. I'll make sure everything's up to date 😁

tranhl commented on .github/workflows/check-dist.yml:
```diff
@@ -0,0 +1,51 @@
+      - name: Rebuild the index.js file
+        run: npm run build
+
+      - name: Compare the expected and actual dist/ directories
```
Full credit to the GitHub actions team here, I've simply based this action from their template!

tranhl commented on .github/workflows/check-dist.yml:
```diff
@@ -0,0 +1,51 @@
+
+      - name: Compare the expected and actual dist/ directories
+        run: |
+          if [ "$(git diff --ignore-space-at-eol dist/ | wc -l)" -gt "0" ]; then
```
Full disclaimer this wasn't written by me, but definitely good to know nonetheless!

tranhl commented on .github/workflows/check-dist.yml:
```diff
@@ -0,0 +1,51 @@
+          node-version: 20.x
+
+      - name: Install dependencies
+        run: npm install --frozen-lockfile
```
Yep definitely makes sense to use `npm ci` instead!

tranhl commented on .github/workflows/check-dist.yml:
```diff
@@ -0,0 +1,51 @@
+        if: ${{ failure() && steps.diff.conclusion == 'failure' }}
+        with:
+          name: dist
+          path: dist/
```
For sure, let me suss that out! I think running Prettier pre-commit and during CI should help with this as well!

tranhl commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -0,0 +1,58 @@
+# You may wish to alter this file to override the set of languages analyzed,
+# or to provide custom queries or build logic.
+#
+# ******** NOTE ********
+# We have attempted to detect the languages in your repository. Please check
+# the `language` matrix defined below to confirm you have the correct set of
+# supported CodeQL languages.
+#
```
Good catch, will zap this as well.

tranhl commented on .github/workflows/git-town.yml:
```diff
@@ -0,0 +1,20 @@
+
+    steps:
+      - uses: actions/checkout@v4
+      - name: Git Town
+        uses: ./
```
Yep that's correct, it uses the local code base. The action doesn't need to build the `dist` folder as the `dist` folder should always contain the built JS bundle equivalent of the source code. That's why we have the `check-dist` workflow, to ensure that any PRs coming in have built the contributed source to `dist/index.js`.

kevgo replied:
Right, I totally forgot that the `dist` folder is checked into the repo even though I was excited about it just a couple of files ago! 🤣 

Wouldn't you need to run `npm ci` here to have the `node_modules` folder populated? That one is also needed but not checked into the repo, no?

tranhl replied:
No `npm ci` required, `esbuild` bundles all of the dependencies into `dist/index.js`!

kevgo replied:
Wow, nice! I didn't know that!

tranhl commented on README.md:
```diff
@@ -1,2 +1,185 @@
+
+It is designed to work out of the box with [Git Town](https://github.com/git-town/git-town) v12+, but also supports previous versions via the `perennial-branches` & `perennial-regex` inputs.
+
+# What's New
```
No reason in particular, I adapted our README from the checkout action's README for familiarity. Happy to increase the heading level for all other headings!

tranhl commented on README.md:
```diff
@@ -1,2 +1,185 @@
+
+Please refer to the [release page](https://github.com/git-town/action/releases/latest) for the latest release notes.
+
+# Usage
```
Sure!

kevgo replied:
Sorry for being unclear. My point is that the "Usage" and "Getting Started" sections contain content that belongs together. Because to use this action, one needs to configure it. So configuring this action is a part of setting it up. Hence there should be only one section that explains how to set up and configure this action, not two. 

The end result could look something like this:

```md
## Installation

### Create the GitHub Actions workflow file

### Create the configuration file for this action

### Configuring perennial branches
```

tranhl replied:
I'm following you now! Let me get that fixed up real quick!

tranhl commented on README.md:
```diff
@@ -1,2 +1,185 @@
+
+jobs:
+  git-town:
+    name: Sync Visualization
```
Done!

tranhl commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -0,0 +1,58 @@
+        # Prefix the list here with "+" to use these queries and those in the config file.
+        # queries: ./path/to/local/query, your-org/your-repo/queries@main
+
+    - run: npm ci
+    - run: npm run build
+    - run: rm -rf dist # We want code scanning to analyze lib instead (individual .js files)
```
Looking into the [run](https://github.com/git-town/action/actions/runs/8399477762/job/23005622978) itself, it seems like CodeQL is taking up the majority of the time. **36s** to initialise CodeQL, and then **42s** to perform code scanning, which I don't think we can avoid? I think that overall a 1 min runtime should be okay, since most PR review will happen asynchronously.

We can definitely remove the redundant install/build step though!

tranhl commented on README.md:
```diff
@@ -1,2 +1,185 @@
+    # The perennial branches for your repository.
+    #
+    # Only necessary when not using `.git-branches.toml` to configure perennial branches.
+    # If perennial branches are specified both in `.git-branches.toml` and this input,
+    # the action will prefer `.git-branches.toml` instead.
+    perennial-branches: |
```
This is a great suggestion, I definitely like this more! And it aligns with how Git config works too. There's a few things we'll need to change:

**Perennial branches**
- Default to no branches
- Apply explicit and regex branches from config
- Apply explicit and regex branches from inputs


**Main branch** 
- Default to default branch from repository
- Apply main branch from config
- Apply main branch from inputs

tranhl commented:
@kevgo All PR feedback has been addressed. I think we're on the final stretch! Let me know if there's anything else your eagle eye spots. Once I get the ✅ I'll go ahead and publish v1 to the Actions Marketplace!

kevgo commented on .github/workflows/check-dist.yml:
```diff
@@ -0,0 +1,70 @@
+        id: build
+        run: npm run build
+
+      # This will fail the workflow if the PR wasn't created by Dependabot.
```
Why should the workflow fail if the PR wasn't created by Dependabot?

tranhl replied:
I'm just as confused as you to be honest 🤷‍♂️. I took the action directly from the [TypeScript Template](https://github.com/actions/typescript-action/blob/1da1fe9abb959743f36210ae5423e1dccc805e85/.github/workflows/check-dist.yml), however nothing in the template or workflow backs up the comment. Perhaps there's some implicit Dependabot behaviour that we're not aware of. Either way, it's confusing probably not worth looking into since the workflow itself works perfectly. I'll remove this comment and the one below.

kevgo commented on .github/workflows/check-dist.yml:
```diff
@@ -0,0 +1,70 @@
+            exit 1
+          fi
+
+      # If `dist/` was different than expected, and this was not a Dependabot
+      # PR, upload the expected version as a workflow artifact.
```
Same here, why the reference to Dependabot?

kevgo commented on .github/workflows/git-town.yml:
```diff
@@ -0,0 +1,24 @@
+name: Git Town
```
Not sure `git-town.yml` is the most appropriate name for this file. How do you feel naming this file `.github/workflows/branch-stack.yml`?

tranhl replied:
I had several motivations for the naming:

- Calling it Git Town would increase exposure, and hopefully by extension adoption. It's similar to how other tools in the ecosystem name their config files by the tool name (e.g. ESLint, Prettier, Vite, etc).
- We could potentially have more functionality we'd like to support using GitHub Actions in the future. If that was the case, adoption could be as simple as adding another job to the workflow file, since we can use descriptive names for jobs within the workflow.

IMO that outweighs the (understandably!) ambiguous name of the workflow file. What are your thoughts?

kevgo replied:
Makes sense to me, thanks for the background!

kevgo commented on .github/workflows/git-town.yml:
```diff
@@ -0,0 +1,24 @@
+  cancel-in-progress: true
+
+jobs:
+  git-town:
```
Adjust this name to whatever you decide to call the file.

tranhl replied:
Regardless of what we agree to in the previous conversation, I do agree that the name for this job could be improved! Will rename to `branch-stack` to match.

kevgo commented on .github/workflows/linting.yml:
```diff
@@ -0,0 +1,63 @@
+        run: |
+          npm exec eslint --max-warnings=0
+
+  editorconfig:
```
Can you simplify this by merging these two jobs (`eslint` and `editorconfig`)? They both do pretty laborious things like installing Node (5-6s) and running `npm ci` (3-4s). The only difference is in the `Lint` steps, which take only 1s in both jobs. 

I propose doing only one job that - as before - installs Node, runs `npm ci`, and then has two `Lint` steps:

```yml
      - name: ESLint
        run: npm exec eslint --max-warnings=0

      - name: Lint .editorconfig
        run: npm exec ec -- -config .editorconfig-checker
```

This would run only 1s slower and would be easier to maintain (something we want to optimize for here).

tranhl replied:
The reason they're two separate jobs is so that even if one job fails, the other can still report its success. If for example, I create a PR that fails both checks, I would fix the issues reported by the first check, only to find out that there were issues with the second. By separating each check into it's own job, we'd only have to fix issues caught by checks once. My team and I adopt this approach at work and it works pretty well for us, especially when we have 10+ checks that run in parallel.

We could create a shared action that encapsulates the "Checkout", "Setup Node", and "Install Dependencies" steps into a single action though! That would de-duplicate the boilerplate leading up to running each check.

kevgo replied:
Yes, that's a pretty common problem! In my experience, if a CI job fails, I have to look at its CLI output anyways to see what exactly failed: which file, which line number, what's the exact error message, what else did go wrong. And when I fix something that broke on CI, I want to verify my fix locally instead of pushing to CI and waiting to see the results there. The latter is too slow.

I therefore usually make it super easy for developers to run the exact same tests locally that CI runs. Or -- even better -- have CI run the same test scripts that developers run. On Git Town there is one [Makefile](https://github.com/git-town/git-town/blob/main/Makefile) that both developers and CI call. For example, when CI runs unit tests, it calls [make unit-race](https://github.com/git-town/git-town/blob/28fa20a4e6f6821191d9b627946f3b50a36bbe33/.github/workflows/unit.yml#L17) the same way I as a developer run unit tests. It runs the same end-to-end tests that developers run because like developers it simply executes [make cuke](https://github.com/git-town/git-town/blob/28fa20a4e6f6821191d9b627946f3b50a36bbe33/.github/workflows/cuke.yml#L17-L18) to run them. When linting, CI also just runs [make fix and make docs](https://github.com/git-town/git-town/blob/28fa20a4e6f6821191d9b627946f3b50a36bbe33/.github/workflows/lint.yml#L17-L20) just as developers do.

So, when CI fails, I just run these tests/linters locally and I'll see the exact same errors and can verify their fixes.

My take is that both checks in your setup are so quick (1 sec each) that running both via a hypothetical `npm run lint` would not be much slower. But it would be much easier than looking up whether the `ESLint` or `Lint .editorconfig` job failed, and then running `npm exec eslint --max-warnings=0` or `npm exec ec -- -config .editorconfig-checker` locally.

Having said all this, I don't feel super strong about this. If you are used to this setup, then it's okay to leave it. Your setup is pretty nice as well and has its advantages.

tranhl replied:
I see where you're coming from! I can see how it would be a lot easier if we made it easier for contributors to run CI checks locally. I'll take a note of it and come back around to it once I have a bit more time to make changes to the CI some more. Thanks for taking the time to explain your approach!

tranhl replied:
Created issue to track: https://github.com/git-town/action/issues/2

kevgo commented on README.md:
```diff
@@ -1,2 +1,151 @@
+:exclamation: Be careful not to add any content between the tag and the
+visualization, as this action will replace that content each time it
+updates your PR. Adding content above the tag, or below the list is
+safe though!
```
Reading this documentation, does it make sense to only insert the stack visualization if there is more than one branch in it? In other words, does it make sense to not show the branch visualization if there would be only one branch in it?

tranhl replied:
When I first introduced this action with my team, this is what it initially did. However the feedback I got was that this was confusing. People weren't sure whether there was an issue with their PR or not. This was compounded by the fact that stacked diffs was a completely different mental model for them as well. Following that feedback, I changed it so that the action always added the visualisation, and that worked out better for my team.

Admittedly, this might not be the case for other users and teams. Perhaps we could run with this for now and gather feedback from the community? I'm more than happy to add support for this (perhaps via an input), but first and foremost I'm really keen to get this in the hands of our citizens so then we can start gathering feedback.

What are your thoughts on this?

kevgo replied:
I think if there is no stack, it could show placeholder text like:

_[not a stacked change]_


But you have more experience with how this feels when being used in real life, so I support whatever you decide here.

tranhl replied:
Ah very true, we could definitely display a placeholder until there is an actual stack! I'd be interested to see how other tools (e.g. Graphite) do this. I'll hold off on any more changes for now to prioritise getting this merged in & released, but definitely open minded about this and down to experiment!

tranhl replied:
Tracked in https://github.com/git-town/action/issues/4

kevgo commented on README.md:
````diff
@@ -1,2 +1,151 @@
+```yaml
+- uses: git-town/action@v1
+  with:
+    main-branch: 'main'
````
Brilliant idea to use inputs for this!

kevgo commented on README.md:
````diff
@@ -1,2 +1,151 @@
+    perennial-regex: '^release-.*$'
+```
+
+Both inputs can be used at the same time. This action will merge the perennial
+branches given into a single, de-duplicated.
````
```suggestion
Both inputs can be used at the same time. This action will merge the perennial
branches given into a single, de-duplicated list.
```

kevgo commented on action.yml:
```diff
@@ -0,0 +1,26 @@
+  perennial-branches:
+    required: false
+    default: ''
+  template:
+    required: false
+    default: |
+      ## Stack
+      {{ stack }}
```
Is this still used? I don't see it referenced anywhere.

tranhl replied:
No! Good catch. I meant to remove the `template` input but removed `perennial-regex` instead. Don't code at 1am in the morning!

kevgo commented on package.json:
```diff
@@ -0,0 +1,37 @@
+    "prettier": "^3.2.5",
+    "remark": "^15.0.1",
+    "remark-gfm": "^4.0.0",
+    "toml": "^3.0.0",
```
Should this be a `dependency`? It is listed as a `devDependency` right now. Separating the two types of dependencies would allow us to use `npm ci --production` on CI. That's significantly snappier than `npm ci`.

tranhl replied:
It would be only a semantic difference, as the dependencies are bundled by `esbuild` prior to the action running, removing the need for `npm ci --production` when running the action.

I do agree though that separating out the dependencies makes it clearer what is a runtime dependency, and what is tooling/dev related. Unfortunately there's no way to install only dev dependencies, so we can't optimise installs for checks that don't need dependencies to be installed.

kevgo replied:
Interesting, thanks for the background. I have to get used to this esbuild bundle thing.

kevgo reviewed (commented):
This is a really nice solution now! I have really only one major suggestion around (optionally) omitting the stack visualization when there is only one branch in it.

tranhl commented:
@

tranhl commented:
@kevgo Ready for another round of review 😊. To summarise what I've changed:

- Removed the confusing mention of Dependabot in the `check-dist` workflow
- Renamed the job ID of the branch stack action to `branch-stack`
- Created a reusable action for setting up the environment for each job, called `setup-env`. This is now called right after `actions/checkout`
- Renamed the `linting.yml` workflow to `ci.yml`, and added a job to run unit tests
- Separated the action's dependencies from `devDependencies` to `dependencies`
- Removed the `template` action input from `action.yml` as it's no longer used. Turns out I mistakenly removed the `perennial-regex` input too, which I've restored
- Used an [alert](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts) instead of an emoji to warn users about adding content between the anchor tag and visualization
- Fixed a typo that you pointed out

There's a couple of review conversations that are still unresolved, as they require further input from you. Keen to hear your thoughts on those! I'm super excited at how things are coming along as well, we're very close I think! 🚀

kevgo commented on action.yml:
```diff
@@ -0,0 +1,24 @@
+name: Git Town - GitHub Action
+description: Visualizes your stacked changes when proposing pull requests on GitHub
```
This description is important for finding this action, let's spend some time word-smithing it. What do you think about this:

```suggestion
description: Adds a map of the corresponding stacked change to pull requests
```

tranhl replied:
Have moved this conversation to this discussion: https://github.com/git-town/action/discussions/3

kevgo reviewed (approved):
I think this is ready to go. I have only nitpicks left 🙂

tranhl commented on .github/actions/setup-env/action.yml:
```diff
@@ -0,0 +1,17 @@
+    - name: Install Dependencies
+      id: install
+      shell: bash
+      run: npm ci
```
Happy to explain! That's correct, `esbuild` will bundle all of an app's dependencies into a single JS file. It also tree-shakes dependencies so if there's any code that you don't use, that will not make it into the final bundle too.

However all of our CI checks operate on the source code, so dependencies need to be installed so that the checks run correctly. For example, ESLint will also perform basic import validation and type checking, which requires that dependencies installed. Also, since we're running tests directly using the source code, dependencies must be installed so that they can resolve correctly inside of the test runtime. This is unique to the way JS operates so I can imagine it'd be different to Go.

Hope that helps explain things!

kevgo replied:
Thanks for the background! According to https://github.com/evanw/esbuild/issues/2713 ESBuild includes any dependency that is used, independent of whether it's listed as `dependency` or `devDependency` in `package.json`. So, if we keep it clean, the JS bundle of this action does not contain the devDependencies since they aren't used by our "production" code. TIL

tranhl commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,3 @@
+## Stack
```
Yep, just so that we can have a header for the visualisation. I do plan to eventually include checklists/additional instructions in the PR template as well to aid contributors.

tranhl commented on README.md:
```diff
@@ -1,2 +1,149 @@
+
+This action visualizes your stacked changes when proposing pull requests on GitHub:
+
+- `main`
+  - ["git town park" command #3181](https://github.com/git-town/git-town/pull/3181) :point_left:
+    - ["git town observe" command" #3186](https://github.com/git-town/git-town/pull/3186)
```
For sure! I'm not a fan of the current iteration either. Unfortunately GitHub doesn't auto link PRs, so I think I'll just add an image instead.
